src/TTImp/Elab/App.idr

```diff
@@ -330,6 +335,20 @@ mutual
                       else pure $ dotTerm tm
            else pure tm
     where
+      
+      ||| Find the constructors of a normal form that is a fully
+      ||| applied type constructor
+      possibleConstructors : NF vars -> Core (List Name)
+      possibleConstructors (NApp _ (NRef (TyCon _ n) tycName) args) = 
+        if length args == n
+        then do defs <- get Ctxt
+                Just gdef <- lookupCtxtExact tycName (gamma defs)
+                | Nothing => pure []
+                let (TCon _ _ _ _ _ _ datacons _) = gdef.definition                                                   | _ => pure []
+                pure datacons
+        else pure []
+      possibleConstructors _ = pure []
```

Comment:
...and this is the other reason (possibly also the line above). If you have a type which is anything other than a `NTCon`, you can't assume there's only one constructor, because it might be matching on a `Type` or some other primitive, e.g. `Int`.

I think otherwise this is a good way to resolve this issue, just that it needs to be careful to only allow things through which are concretely in `NTCon` form.

Reply:
OK, I'll fix. There's another bug: using the empty list as a failure is exactly the wrong result to return.
I'll use a `Maybe` instead.